File: source/extensions/omni.isaac.sensor/python/nodes/OgnIsaacReadContactSensor.py
# Copyright (c) 2022-2023, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto. Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.
#
# Array or tuple values are accessed as numpy arrays so you probably need this import
import numpy
import omni.graph.core as og
import omni.timeline
import usdrt.Sdf
from omni.isaac.core_nodes import BaseResetNode
from omni.isaac.dynamic_control import _dynamic_control
from omni.isaac.sensor import _sensor
from omni.isaac.sensor.ogn.OgnIsaacReadContactSensorDatabase import OgnIsaacReadContactSensorDatabase


class OgnIsaacReadContactSensorInternalState(BaseResetNode):
    def __init__(self):
        self._cs = _sensor.acquire_contact_sensor_interface()
        self.cs_path = ""
        super().__init__(initialize=False)

    def custom_reset(self):
        self.cs_path = ""
        pass


class OgnIsaacReadContactSensor:
    """
    Node that returns Contact Sensor data
    """

    @staticmethod
    def internal_state():
        return OgnIsaacReadContactSensorInternalState()

    @staticmethod
    def compute(db) -> bool:
        state = db.internal_state

        if len(db.inputs.csPrim) > 0:
            state.cs_path = db.inputs.csPrim[0].GetString()

        else:
            db.outputs.inContact = False
            db.outputs.value = 0
            db.outputs.sensorTime = 0
            db.log_error("Invalid contact sensor prim")
            return False

        reading = state._cs.get_sensor_reading(state.cs_path)
        if reading.is_valid:
            db.outputs.inContact = reading.in_contact
            db.outputs.value = reading.value
            db.outputs.sensorTime = reading.time
        else:
            db.outputs.inContact = False
            db.outputs.value = 0
            db.outputs.sensorTime = 0
            db.log_warn("Invalid contact sensor measurement, is it enabled?")
            return False

        db.outputs.execOut = og.ExecutionAttributeState.ENABLED

        return True

    @staticmethod
    def release(node):
        try:
            state = OgnIsaacReadContactSensorDatabase.per_node_internal_state(node)
        except Exception:
            state = None
            pass

        if state is not None:
            state.reset()

File: source/extensions/omni.isaac.sensor/nodes/OgnIsaacReadContactSensor.cpp
// Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
//
// NVIDIA CORPORATION and its licensors retain all intellectual property
// and proprietary rights in and to this software, related documentation
// and any modifications thereto. Any use, reproduction, disclosure or
// distribution of this software and related documentation without an express
// license agreement from NVIDIA CORPORATION is strictly prohibited.
//

// clang-format off
#include <pch/UsdPCH.h>
// clang-format on

#include "omni/isaac/utils/UsdUtilities.h"

#include <omni/fabric/FabricUSD.h>

#include <IsaacSensor.h>
#include <OgnIsaacReadContactSensorDatabase.h>


namespace omni
{
namespace isaac
{
namespace sensor
{

class OgnIsaacReadContactSensor
{
public:
    static void initInstance(NodeObj const& nodeObj, GraphInstanceID instanceId)
    {
        auto& state =
            OgnIsaacReadContactSensorDatabase::sPerInstanceState<OgnIsaacReadContactSensor>(nodeObj, instanceId);

        state.mContactSensorInterface = carb::getCachedInterface<ContactSensorInterface>();

        if (!state.mContactSensorInterface)
        {
            CARB_LOG_ERROR("Failed to acquire omni::isaac::sensor interface");
            return;
        }
    }

    static bool compute(OgnIsaacReadContactSensorDatabase& db)
    {
        auto& state = db.perInstanceState<OgnIsaacReadContactSensor>();

        const auto& prim = db.inputs.csPrim();
        const char* primPath;

        auto& inContact = db.outputs.inContact();
        auto& value = db.outputs.value();
        auto& sensorTime = db.outputs.sensorTime();

        if (prim.size() > 0)
        {
            primPath = omni::fabric::toSdfPath(prim[0]).GetText();
        }
        else
        {
            inContact = false;
            value = 0;
            sensorTime = 0.0f;
            db.logError("Invalid contact sensor prim");
            return false;
        }

        CsReading sensorReading = state.mContactSensorInterface->getSensorReading(primPath, db.inputs.useLatestData());

        if (sensorReading.is_valid)
        {
            inContact = sensorReading.inContact;
            value = sensorReading.value;
            sensorTime = sensorReading.time;
        }
        else
        {
            inContact = false;
            value = 0;
            sensorTime = 0.0f;
            db.logWarning("no valid sensor reading, is the sensor enabled?");
            db.outputs.execOut() = kExecutionAttributeStateDisabled;
            return false;
        }
        db.outputs.execOut() = kExecutionAttributeStateEnabled;
        return true;
    }

private:
    ContactSensorInterface* mContactSensorInterface = nullptr;
};


REGISTER_OGN_NODE()
} // sensor
} // graph
} // omni
